[Observability AI Assistant] More specific error handling

dgieselaar opened the pull request:
Implements more specific error handling, in addition to a bug fix w/ the Lens function where it was not rendering outside of the Observability AI Assistant plugin's context, and validation of the parameters before executing the function.

Additionally:
- improves recall function
- filter alerts and services by alert/health status

<!--ONMERGE {"backportTargets":["8.10"]} ONMERGE-->

dgieselaar committed:
[Observability AI Assistant] More specific error handling

Implements more specific error handling, in addition to a bug fix w/ the Lens function where it was not rendering outside of the Observability AI Assistant plugin's context, and validation of the parameters before executing the function.

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

sqren commented on package.json:
```diff
@@ -90,6 +90,7 @@
   "dependencies": {
     "@appland/sql-parser": "^1.5.1",
     "@babel/runtime": "^7.21.0",
+    "@cfworker/json-schema": "^1.12.7",
```
Iisn't this meant to run on cloudflare workers?

wayneseymour commented on x-pack/plugins/apm/server/routes/assistant_functions/route.ts:
```diff
@@ -253,7 +263,7 @@ const getApmServicesListRoute = createApmServerRoute({
     });
 
-    const mappedItems = serviceItems.items.map((item): ApmServicesListItem => {
+    let mappedItems = serviceItems.items.map((item): ApmServicesListItem => {
```
curious.  why `let` instead of `const` here?

dgieselaar replied:
we re-assign on line 278

sqren replied:
This also caught my eye. No biggie, but `let` requires the reviewer to read ahead, so for that reason I prefer `const`.

You can avoid re-assign by unconditionally filtering:


```json
  .filter((item) => {
    if (healthStatus && healthStatus.length) {
      return healthStatus.includes(item.healthStatus);
    }

    return true;
  });    
```

dgieselaar replied:
I'd usually agree but I plan to add more filters here, where a let is more appropriate I think

dgieselaar commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into obs-ai-assistant-error-handling

dgieselaar committed:
[Observability AI Assistant] More specific error handling

Implements more specific error handling, in addition to a bug fix w/ the Lens function where it was not rendering outside of the Observability AI Assistant plugin's context, and validation of the parameters before executing the function.

sqren commented on x-pack/plugins/apm/server/routes/assistant_functions/route.ts:
```diff
@@ -264,6 +274,12 @@ const getApmServicesListRoute = createApmServerRoute({
       };
     });
 
+    if (healthStatus && healthStatus.length) {
+      mappedItems = mappedItems.filter((item): boolean =>
+        healthStatus.includes(item.healthStatus)
+      );
+    }
+
```
nit: you can avoid re-assigning 

```suggestion
    })
  .filter((item) => {
    if (healthStatus && healthStatus.length) {
      return healthStatus.includes(item.healthStatus);
    }

    return true;
  });    
```

dgieselaar committed:
Merge branch 'obs-ai-assistant-error-handling' of github.com:dgieselaar/kibana into obs-ai-assistant-error-handling

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.10|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 165068
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->

dgieselaar commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.10|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/165188">](https://github.com/elastic/kibana/pull/165188)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->
